java-checks/src/main/java/org/sonar/java/checks/SwitchCaseWithoutBreakCheck.java

```diff
@@ -19,75 +19,82 @@
  */
 package org.sonar.java.checks;
 
-import com.google.common.collect.Iterables;
+import com.google.common.collect.Lists;
 import org.sonar.check.Rule;
 import org.sonar.plugins.java.api.JavaFileScanner;
 import org.sonar.plugins.java.api.JavaFileScannerContext;
 import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
-import org.sonar.plugins.java.api.tree.BreakStatementTree;
-import org.sonar.plugins.java.api.tree.CaseGroupTree;
-import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.BlockTree;
+import org.sonar.plugins.java.api.tree.CaseLabelTree;
+import org.sonar.plugins.java.api.tree.IfStatementTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
 import org.sonar.plugins.java.api.tree.SwitchStatementTree;
-import org.sonar.plugins.java.api.tree.ThrowStatementTree;
-
-import java.util.ArrayDeque;
-import java.util.Deque;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TryStatementTree;
 
 @Rule(key = "S128")
 public class SwitchCaseWithoutBreakCheck extends BaseTreeVisitor implements JavaFileScanner {
 
   private JavaFileScannerContext context;
 
-  private final Deque<CaseGroupTree> invalidCaseGroups = new ArrayDeque<>();
-  private CaseGroupTree currentTree = null;
-
   @Override
   public void scanFile(JavaFileScannerContext context) {
     this.context = context;
     scan(context.getTree());
   }
 
   @Override
-  public void visitSwitchStatement(SwitchStatementTree tree) {
-    scan(tree.expression());
-    if (!tree.cases().isEmpty()){
-      scan(tree.cases().subList(0, tree.cases().size() - 1));
-    }
-  }
+  public void visitSwitchStatement(SwitchStatementTree switchStatement) {
+      switchStatement.cases().stream()
+          // Exclude the last case as stated in RSPEC. This also excludes switches with no or a single case group.
+          .limit(Math.max(0, switchStatement.cases().size() - 1))
+          .forEach(caseGroup -> {
+            // Assign issues to the last label in the group
+            CaseLabelTree caseLabel = caseGroup.labels().get(caseGroup.labels().size() - 1);
 
-  @Override
-  public void visitCaseGroup(CaseGroupTree tree) {
-    currentTree = tree;
-    invalidCaseGroups.push(tree);
+            // Reverse the body as commonly the unconditional exit will be at the end of the body.
+            if (Lists.reverse(caseGroup.body()).stream().noneMatch(SwitchCaseWithoutBreakCheck::isUnconditionalExit)) {
+              context.reportIssue(this, caseLabel, "End this switch case with an unconditional break, return or throw statement.");
+            }
+          });
 
-    super.visitCaseGroup(tree);
+    super.visitSwitchStatement(switchStatement);
+  }
 
-    if (invalidCaseGroups.remove(tree)) {
-      context.reportIssue(this, Iterables.getLast(tree.labels()), "End this switch case with an unconditional break, return or throw statement.");
+  private static boolean isUnconditionalExit(Tree syntaxNode) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "isUnconditionalExit" is 13 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)